File: skills/spec-driven-development/scripts/validate_spec.py
#!/usr/bin/env python3
"""Validate a Markdown SDD specification for minimum planning readiness."""

from __future__ import annotations

import argparse
import json
import re
import sys
from dataclasses import asdict
from dataclasses import dataclass
from pathlib import Path


SPEC_ID_RE = re.compile(r"\b(?:REQ|API|SCHEMA|STATE|ERR|AC|NFR|MIG|OBS|NON)-\d{3,}\b")
PLACEHOLDER_RE = re.compile(r"<[^>\n]{2,}>|\bTODO\b|\bTBD\b|待补充|待定|占位")
MAYBE_AMBIGUOUS = ("适当", "友好", "常见情况", "尽快", "合理", "必要时")
ALLOWED_TRACE_STATUSES = {"planned", "covered", "manual", "blocked", "deferred", "not-applicable", "n/a"}


@dataclass(frozen=True)
class Table:
    headers: list[str]
    rows: list[list[str]]
    start_line: int


@dataclass(frozen=True)
class ValidationResult:
    path: str
    ok: bool
    errors: list[str]
    warnings: list[str]


def parse_args(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Validate a Markdown SDD spec for Spec IDs, traceability, and placeholders."
    )
    parser.add_argument("spec_files", nargs="+", help="Path(s) to Markdown specification files.")
    parser.add_argument(
        "--format",
        choices=("text", "json"),
        default="text",
        help="Output format. Use json for CI or editor integrations.",
    )
    parser.add_argument(
        "--strict",
        action="store_true",
        help="Treat warning-level ambiguous wording as validation errors.",
    )
    return parser.parse_args(argv)


def split_table_row(line: str) -> list[str]:
    return [cell.strip() for cell in line.strip().strip("|").split("|")]


def is_separator_row(cells: list[str]) -> bool:
    return bool(cells) and all(re.fullmatch(r":?-{3,}:?", cell.strip()) for cell in cells)


def extract_tables(lines: list[str]) -> list[Table]:
    tables: list[Table] = []
    i = 0
    while i < len(lines):
        if not lines[i].lstrip().startswith("|") or i + 1 >= len(lines):
            i += 1
            continue

        headers = split_table_row(lines[i])
        separator = split_table_row(lines[i + 1])
        if not is_separator_row(separator):
            i += 1
            continue

        rows: list[list[str]] = []
        start_line = i + 1
        i += 2
        while i < len(lines) and lines[i].lstrip().startswith("|"):
            row = split_table_row(lines[i])
            if not is_separator_row(row):
                rows.append(row)
            i += 1
        tables.append(Table(headers=headers, rows=rows, start_line=start_line))
    return tables


def get_cell(row: list[str], headers: list[str], names: tuple[str, ...]) -> str:
    normalized = [header.strip().lower() for header in headers]
    for name in names:
        name = name.lower()
        for index, header in enumerate(normalized):
            if name in header and index < len(row):
                return row[index].strip()
    return ""


def contains_placeholder(text: str) -> bool:
    return bool(PLACEHOLDER_RE.search(text))


def normalize_cell(text: str) -> str:
    return text.strip().strip("`").strip()


def validate_spec(path: Path, strict: bool) -> tuple[list[str], list[str]]:
    errors: list[str] = []
    warnings: list[str] = []

    if not path.exists():
        return [f"File does not exist: {path}"], warnings
    if not path.is_file():
        return [f"Path is not a file: {path}"], warnings

    text = path.read_text(encoding="utf-8")
    lines = text.splitlines()
    if not text.strip():
        return ["Spec file is empty."], warnings

    for match in PLACEHOLDER_RE.finditer(text):
        line_no = text.count("\n", 0, match.start()) + 1
        errors.append(f"Line {line_no}: unresolved placeholder or TODO: {match.group(0)!r}.")

    for phrase in MAYBE_AMBIGUOUS:
        if phrase in text:
            warnings.append(f"Ambiguous wording found: {phrase!r}. Make it observable if it is normative.")

    ids = set(SPEC_ID_RE.findall(text))
    if not ids:
        errors.append("No Spec ID found. Use REQ/API/SCHEMA/STATE/ERR/AC/NFR/MIG/OBS/NON-001 style IDs.")

    tables = extract_tables(lines)
    must_ids: set[str] = set()
    should_ids: set[str] = set()
    trace_ids: set[str] = set()
    defined_ids: dict[str, int] = {}

    for table in tables:
        header_text = " | ".join(table.headers).lower()
        is_trace_table = "spec id" in header_text and (
            "verification" in header_text or "test file" in header_text or "command" in header_text
        )
        has_status_column = any("status" in header.strip().lower() for header in table.headers)

        for offset, row in enumerate(table.rows, start=2):
            line_no = table.start_line + offset
            row_text = " | ".join(row)
            row_ids = set(SPEC_ID_RE.findall(row_text))
            priority = normalize_cell(get_cell(row, table.headers, ("priority",))).upper()

            if row_ids and not is_trace_table:
                for spec_id in row_ids:
                    if spec_id in defined_ids:
                        errors.append(
                            f"Line {line_no}: Duplicate Spec ID definition {spec_id} "
                            f"(first defined on line {defined_ids[spec_id]})."
                        )
                    else:
                        defined_ids[spec_id] = line_no

            if priority in {"MUST", "SHOULD"} and not row_ids:
                errors.append(f"Line {line_no}: {priority} row is missing a Spec ID.")

            if priority == "MUST":
                must_ids.update(row_ids)
                verification = get_cell(row, table.headers, ("verification",))
                if not verification or contains_placeholder(verification):
                    errors.append(f"Line {line_no}: MUST row lacks concrete verification evidence.")

            if priority == "SHOULD":
                should_ids.update(row_ids)

            if is_trace_table and row_ids:
                trace_ids.update(row_ids)
                verification_type = get_cell(row, table.headers, ("verification type", "verification"))
                test_command = get_cell(row, table.headers, ("test file", "command", "evidence"))
                if (
                    not verification_type
                    or not test_command
                    or contains_placeholder(verification_type)
                    or contains_placeholder(test_command)
                ):
                    errors.append(f"Line {line_no}: traceability row for {sorted(row_ids)} lacks evidence.")
                if has_status_column:
                    status = get_cell(row, table.headers, ("status",))
                    normalized_status = normalize_cell(status).lower()
                    if not status or contains_placeholder(status):
                        errors.append(f"Line {line_no}: traceability row for {sorted(row_ids)} lacks status.")
                    elif normalized_status not in ALLOWED_TRACE_STATUSES:
                        allowed = ", ".join(sorted(ALLOWED_TRACE_STATUSES))
                        errors.append(
                            f"Line {line_no}: invalid traceability status {status!r}; allowed: {allowed}."
                        )

    if must_ids and not trace_ids:
        errors.append("MUST requirements exist but no Traceability Matrix rows were found.")

    missing_trace = sorted(must_ids - trace_ids)
    if missing_trace:
        errors.append("MUST requirements missing from Traceability Matrix: " + ", ".join(missing_trace) + ".")

    missing_should_trace = sorted(should_ids - trace_ids)
    if missing_should_trace:
        warnings.append(
            "SHOULD requirements missing from Traceability Matrix: " + ", ".join(missing_should_trace) + "."
        )

    if strict and warnings:
        errors.extend(warnings)
        warnings = []

    return errors, warnings


def build_result(path: Path, strict: bool) -> ValidationResult:
    errors, warnings = validate_spec(path, strict)
    return ValidationResult(path=str(path), ok=not errors, errors=errors, warnings=warnings)


def print_text_results(results: list[ValidationResult]) -> None:
    for index, result in enumerate(results):
        if index:
            print()

        if result.errors:
            print(f"Spec validation failed: {result.path}")
            print("\nErrors:")
            for error in result.errors:
                print(f"- {error}")
        else:
            print(f"Spec validation passed: {result.path}")

        if result.warnings:
            print("\nWarnings:")
            for warning in result.warnings:
                print(f"- {warning}")


def main(argv: list[str]) -> int:
    args = parse_args(argv)
    results = [build_result(Path(spec_file), args.strict) for spec_file in args.spec_files]
    ok = all(result.ok for result in results)

    if args.format == "json":
        payload = {
            "ok": ok,
            "error_count": sum(len(result.errors) for result in results),
            "warning_count": sum(len(result.warnings) for result in results),
            "results": [asdict(result) for result in results],
        }
        print(json.dumps(payload, ensure_ascii=False, indent=2))
    else:
        print_text_results(results)

    return 0 if ok else 1


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: src/lib/validate-spec.ts
import { existsSync, readFileSync, statSync } from "node:fs";

const SPEC_ID_RE = /\b(?:REQ|API|SCHEMA|STATE|ERR|AC|NFR|MIG|OBS|NON)(?:-[A-Z0-9]+)*-\d{3,}\b/g;
const PLACEHOLDER_RE = /<[^>\n]{2,}>|\bTODO\b|\bTBD\b|待补充|待定|占位/g;
const SEPARATOR_RE = /^:?-{3,}:?$/;

const MAYBE_AMBIGUOUS = ["适当", "友好", "常见情况", "尽快", "合理", "必要时"];

const ALLOWED_TRACE_STATUSES = new Set([
  "planned",
  "covered",
  "manual",
  "blocked",
  "deferred",
  "not-applicable",
  "n/a",
  "implemented",
  "done",
  "计划中",
  "已覆盖",
  "已实现",
  "完成",
  "人工",
  "阻塞",
  "延期",
  "不适用",
]);

interface Table {
  headers: string[];
  rows: string[][];
  startLine: number;
}

export interface ValidationResult {
  path: string;
  ok: boolean;
  errors: string[];
  warnings: string[];
}

export interface ValidationPayload {
  ok: boolean;
  error_count: number;
  warning_count: number;
  results: ValidationResult[];
}

function allMatches(regex: RegExp, text: string): RegExpExecArray[] {
  const clone = new RegExp(regex.source, regex.flags);
  const matches: RegExpExecArray[] = [];
  let match: RegExpExecArray | null;
  while ((match = clone.exec(text)) !== null) {
    matches.push(match);
  }
  return matches;
}

function splitTableRow(line: string): string[] {
  return line.trim().replace(/^\|/, "").replace(/\|$/, "").split("|").map((cell) => cell.trim());
}

function isSeparatorRow(cells: string[]): boolean {
  return cells.length > 0 && cells.every((cell) => SEPARATOR_RE.test(cell.trim()));
}

function extractTables(lines: string[]): Table[] {
  const tables: Table[] = [];
  let index = 0;
  while (index < lines.length) {
    if (!lines[index].trimStart().startsWith("|") || index + 1 >= lines.length) {
      index += 1;
      continue;
    }

    const headers = splitTableRow(lines[index]);
    const separator = splitTableRow(lines[index + 1]);
    if (!isSeparatorRow(separator)) {
      index += 1;
      continue;
    }

    const rows: string[][] = [];
    const startLine = index + 1;
    index += 2;
    while (index < lines.length && lines[index].trimStart().startsWith("|")) {
      const row = splitTableRow(lines[index]);
      if (!isSeparatorRow(row)) {
        rows.push(row);
      }
      index += 1;
    }
    tables.push({ headers, rows, startLine });
  }
  return tables;
}

function getCell(row: string[], headers: string[], names: string[]): string {
  const normalized = headers.map((header) => header.trim().toLowerCase());
  for (const rawName of names) {
    const name = rawName.toLowerCase();
    for (const [index, header] of normalized.entries()) {
      if (header.includes(name) && index < row.length) {
        return row[index].trim();
      }
    }
  }
  return "";
}

function hasHeader(headers: string[], names: string[]): boolean {
  return names.some((name) => getCell(headers, headers, [name]));
}

function placeholderMatches(text: string): RegExpExecArray[] {
  return allMatches(PLACEHOLDER_RE, text).filter((match) => {
    const value = match[0];
    if (value.startsWith("<")) {
      const previous = match.index > 0 ? text[match.index - 1] : "";
      if (/[A-Za-z0-9]/.test(previous) || previous === "_" || previous === ">") {
        return false;
      }
    }
    return true;
  });
}

function containsPlaceholder(text: string): boolean {
  return placeholderMatches(text).length > 0;
}

function normalizeCell(text: string): string {
  return text.trim().replace(/^`+|`+$/g, "").trim();
}

function normalizePriority(text: string): string {
  const normalized = normalizeCell(text).toUpperCase();
  if (["必须", "必需", "MUST"].includes(normalized)) {
    return "MUST";
  }
  if (["应该", "SHOULD"].includes(normalized)) {
    return "SHOULD";
  }
  return normalized;
}

function isTraceabilityTable(headers: string[]): boolean {
  const headerText = headers.join(" | ").toLowerCase();
  const hasSpecId = headerText.includes("spec id") || headerText.includes("规格 id") || headerText.includes("规格id");
  const hasEvidence = ["verification", "test file", "command", "evidence", "验证", "测试文件", "命令", "证据"].some(
    (marker) => headerText.includes(marker),
  );
  return hasSpecId && hasEvidence;
}

function sortedDisplay(values: Set<string>): string {
  return `['${[...values].sort().join("', '")}']`;
}

function specIds(text: string): Set<string> {
  return new Set(allMatches(SPEC_ID_RE, text).map((match) => match[0]));
}

export function validateSpec(path: string, strict: boolean): [string[], string[]] {
  const errors: string[] = [];
  let warnings: string[] = [];

  if (!existsSync(path)) {
    return [[`File does not exist: ${path}`], warnings];
  }
  if (!statSync(path).isFile()) {
    return [[`Path is not a file: ${path}`], warnings];
  }

  const text = readFileSync(path, "utf8");
  const lines = text.split(/\r?\n/);
  if (!text.trim()) {
    return [["Spec file is empty."], warnings];
  }

  for (const match of placeholderMatches(text)) {
    const lineNo = text.slice(0, match.index).split("\n").length;
    errors.push(`Line ${lineNo}: unresolved placeholder or TODO: '${match[0]}'.`);
  }

  for (const phrase of MAYBE_AMBIGUOUS) {
    if (text.includes(phrase)) {
      warnings.push(`Ambiguous wording found: '${phrase}'. Make it observable if it is normative.`);
    }
  }

  const ids = specIds(text);
  if (ids.size === 0) {
    errors.push("No Spec ID found. Use REQ/API/SCHEMA/STATE/ERR/AC/NFR/MIG/OBS/NON-001 style IDs.");
  }

  const tables = extractTables(lines);
  const mustIds = new Set<string>();
  const shouldIds = new Set<string>();
  const traceIds = new Set<string>();
  const definedIds = new Map<string, number>();

  for (const table of tables) {
    const isTraceTable = isTraceabilityTable(table.headers);
    const hasStatusColumn = hasHeader(table.headers, ["status", "状态"]);

    table.rows.forEach((row, rowIndex) => {
      const lineNo = table.startLine + rowIndex + 2;
      const rowText = row.join(" | ");
      const rowIds = specIds(rowText);
      const priority = normalizePriority(getCell(row, table.headers, ["priority", "优先级"]));

      if (rowIds.size > 0 && !isTraceTable) {
        for (const specId of rowIds) {
          if (definedIds.has(specId)) {
            errors.push(
              `Line ${lineNo}: Duplicate Spec ID definition ${specId} (first defined on line ${definedIds.get(specId)}).`,
            );
          } else {
            definedIds.set(specId, lineNo);
          }
        }
      }

      if ((priority === "MUST" || priority === "SHOULD") && rowIds.size === 0) {
        errors.push(`Line ${lineNo}: ${priority} row is missing a Spec ID.`);
      }

      if (priority === "MUST") {
        for (const specId of rowIds) {
          mustIds.add(specId);
        }
        const verification = getCell(row, table.headers, ["verification", "验证方式", "验证"]);
        if (!verification || containsPlaceholder(verification)) {
          errors.push(`Line ${lineNo}: MUST row lacks concrete verification evidence.`);
        }
      }

      if (priority === "SHOULD") {
        for (const specId of rowIds) {
          shouldIds.add(specId);
        }
      }

      if (isTraceTable && rowIds.size > 0) {
        for (const specId of rowIds) {
          traceIds.add(specId);
        }
        const verificationType = getCell(row, table.headers, [
          "verification type",
          "verification",
          "验证类型",
          "验证方式",
        ]);
        const testCommand = getCell(row, table.headers, ["test file", "command", "evidence", "测试文件", "命令", "证据"]);
        if (
          !verificationType ||
          !testCommand ||
          containsPlaceholder(verificationType) ||
          containsPlaceholder(testCommand)
        ) {
          errors.push(`Line ${lineNo}: traceability row for ${sortedDisplay(rowIds)} lacks evidence.`);
        }
        if (hasStatusColumn) {
          const status = getCell(row, table.headers, ["status", "状态"]);
          const normalizedStatus = normalizeCell(status).toLowerCase();
          if (!status || containsPlaceholder(status)) {
            errors.push(`Line ${lineNo}: traceability row for ${sortedDisplay(rowIds)} lacks status.`);
          } else if (!ALLOWED_TRACE_STATUSES.has(normalizedStatus)) {
            const allowed = [...ALLOWED_TRACE_STATUSES].sort().join(", ");
            errors.push(`Line ${lineNo}: invalid traceability status '${status}'; allowed: ${allowed}.`);
          }
        }
      }
    });
  }

  if (mustIds.size > 0 && traceIds.size === 0) {
    errors.push("MUST requirements exist but no Traceability Matrix rows were found.");
  }

  const missingTrace = [...mustIds].filter((id) => !traceIds.has(id)).sort();
  if (missingTrace.length > 0) {
    errors.push(`MUST requirements missing from Traceability Matrix: ${missingTrace.join(", ")}.`);
  }

  const missingShouldTrace = [...shouldIds].filter((id) => !traceIds.has(id)).sort();
  if (missingShouldTrace.length > 0) {
    warnings.push(`SHOULD requirements missing from Traceability Matrix: ${missingShouldTrace.join(", ")}.`);
  }

  if (strict && warnings.length > 0) {
    errors.push(...warnings);
    warnings = [];
  }

  return [errors, warnings];
}

export function buildResult(path: string, strict: boolean): ValidationResult {
  const [errors, warnings] = validateSpec(path, strict);
  return { path, ok: errors.length === 0, errors, warnings };
}

export function buildPayload(paths: string[], strict: boolean): ValidationPayload {
  const results = paths.map((path) => buildResult(path, strict));
  return {
    ok: results.every((result) => result.ok),
    error_count: results.reduce((total, result) => total + result.errors.length, 0),
    warning_count: results.reduce((total, result) => total + result.warnings.length, 0),
    results,
  };
}

export function formatTextResults(results: ValidationResult[]): string {
  const blocks = results.map((result) => {
    const lines: string[] = [];
    if (result.errors.length > 0) {
      lines.push(`Spec validation failed: ${result.path}`, "", "Errors:");
      for (const error of result.errors) {
        lines.push(`- ${error}`);
      }
    } else {
      lines.push(`Spec validation passed: ${result.path}`);
    }
    if (result.warnings.length > 0) {
      lines.push("", "Warnings:");
      for (const warning of result.warnings) {
        lines.push(`- ${warning}`);
      }
    }
    return lines.join("\n");
  });
  return blocks.join("\n\n");
}
